common/djangoapps/third_party_auth/provider.py

```diff
@@ -30,19 +31,17 @@ def _enabled_providers(cls):
         Helper method that returns a generator used to iterate over all providers
         of the current site.
         """
-        oauth2_backend_names = OAuth2ProviderConfig.key_values('backend_name', flat=True)
-        for oauth2_backend_name in oauth2_backend_names:
-            provider = OAuth2ProviderConfig.current(oauth2_backend_name)
+        oauth2_backends = OAuth2ProviderConfig.objects.prefetch_related('site').all()
+        for provider in unique_everseen(oauth2_backends, key=lambda p: p.backend_name):
             if provider.enabled_for_current_site and provider.backend_name in _PSA_OAUTH2_BACKENDS:
```

Comment:
Since we're only ever doing this for the site we're on, can we do the reverse and get all the providers where the site is the current site?

If I'm reading this right, it looks like a provider is mapped to at most one site:
https://github.com/edx/edx-platform/blob/0f5c474f8f49e7dbc1a7d60e15f5409a29dfe69f/common/djangoapps/third_party_auth/models.py#L138

Reply:
Agreed that that would make more sense. For that matter, why not add `site`, `enabled`, and `backend_name` to the query and skip the entire guard clause that's inside the for loop?

I was being cautious because I'm not sure why it wasn't written that way in the first place... but maybe there's not a good reason. :-)

Reply:
Yea, if we find one later, we can always undo the change and put in the good reason as a comment.  Let's just make this a more sensible query.

Reply:
Curiously, most of the tests pass if I do the below, including one of the two that was failing before—but the other one (`test_provider.py:38`) is still failing as if the query is being cached.

```py
        site = Site.objects.get_current(get_current_request())
        oauth2_backends = OAuth2ProviderConfig.objects.filter(enabled=True, site=site, backend_name__in=_PSA_OAUTH2_BACKENDS)
        for provider in unique_everseen(oauth2_backends, key=lambda p: p.backend_name):
            yield provider
```

Failure:

```py
    def test_runtime_configuration(self):
        self.configure_google_provider(enabled=True)
        enabled_providers = provider.Registry.enabled()
        self.assertEqual(len(enabled_providers), 1)
        self.assertEqual(enabled_providers[0].name, "Google")
        self.assertEqual(enabled_providers[0].get_setting("SECRET"), "opensesame")
    
        self.configure_google_provider(enabled=False)
        enabled_providers = provider.Registry.enabled()
>       self.assertEqual(len(enabled_providers), 0)
E       AssertionError: 1 != 0

common/djangoapps/third_party_auth/tests/test_provider.py:38: AssertionError
```

Reply:
OK, I think I see what's happening here. The test configures a provider that is enabled, then configures one that is disabled, both with the same backend name, so that we end up with two objects in the DB. The code on master is supposed to pick up _only the most recent object_, and then discard it if it is disabled. That's why it's not part of the query.

(But then I don't understand why site wasn't part of the query! Surely we'd want to partition that append-to-override behavior by site.)

Reply:
(And the reason my code was failing is because it was getting the _first_ row with a given slug or backend or key, not the _last_.)

Reply:
So this is because it's backed by config models right?  The config model table is essentially keeping a history of all the ways a setting has been set. So to get the correct setting, we want to get the latest setting for any given name and treat it as the source of truth.  So if the latest version is disabled, then we treat the value as disabled.  It seems like the test is valid.

Reply:
Yeah, and my iterator was skipping the disabled one. I think the existing code has a bug around handling of sites, though; if two providers are registered for the same `backend_name` but different sites, the existing code makes the more recent one shadow the other.